capsules/src/ros.rs

```diff
@@ -0,0 +1,142 @@
+//! Read Only Systems calls
+//!
+//! This capsule provides read only system calls to userspace applications.
+//! This is similar to the Linux vDSO syscalls.
+//!
+//! The benefit of using these is that applications can avoid the context
+//! switch overhead of traditional syscalls by just reading the value from
+//! memory.
+//!
+//! The value will only be as accurate as the last time the application was
+//! switched to by the kernel.
+//!
+//! The layout of the read only syscalls in the allow region depends on the
+//! version. Userspace can use `command 0` to get the version information.
+//!
+//! Versions are backwards compatible, that is new versions will only add
+//! fields, not remove existing ones or change the order.
+//!
+//! Version 1:
+//!   |-------------------------|
+//!   |       Count (u32)       |
+//!   |-------------------------|
+//!   |   Pending Tasks (u32)   |
+//!   |-------------------------|
+//!   |                         |
+//!   |     Time Ticks (u64)    |
+//!   |-------------------------|
+//!
+
+use core::cell::Cell;
+use kernel::grant::Grant;
+use kernel::hil::time::{Ticks, Time};
+use kernel::platform::ContextSwitchOps;
+use kernel::process::{self, ProcessId};
+use kernel::processbuffer::{SharedProcessBuffer, WriteableProcessBuffer};
+use kernel::syscall::{CommandReturn, SyscallDriver};
+use kernel::ErrorCode;
+
+/// Syscall driver number.
+pub const DRIVER_NUM: usize = 0x10001;
+const VERSION: u32 = 1;
+
+pub struct ROSDriver<'a, T: Time> {
+    timer: &'a T,
+
+    count: Cell<u32>,
+    apps: Grant<App, 0>,
+}
+
+impl<'a, T: Time> ROSDriver<'a, T> {
+    pub fn new(timer: &'a T, grant: Grant<App, 0>) -> ROSDriver<'a, T> {
+        ROSDriver {
+            timer,
+            count: Cell::new(0),
+            apps: grant,
+        }
+    }
+}
+
+impl<'a, T: Time> ContextSwitchOps for ROSDriver<'a, T> {
+    fn update_values(&self, process: &dyn process::Process) {
+        let count = self.count.get();
+        let appid = process.processid();
+        let pending_tasks = process.pending_tasks();
+
+        self.apps
+            .enter(appid, |app, _| {
+                let _ = app.mem_region.mut_enter(|buf| {
+                    if buf.len() >= 4 {
+                        buf[0..4].copy_from_slice(&count.to_le_bytes());
+                    }
+                    if buf.len() >= 8 {
+                        buf[4..8].copy_from_slice(&(pending_tasks as u32).to_le_bytes());
+                    }
+                    if buf.len() >= 16 {
+                        let now = self.timer.now().into_usize() as u64;
+                        buf[8..16].copy_from_slice(&now.to_le_bytes());
+                    }
+                });
+            })
+            .unwrap();
+
+        self.count.set(count.wrapping_add(1));
```

Comment:
Is this a duplicate of the count that's in the process debug info?

Reply:
I suspect so. This is something that @phil-levis wanted

Reply:
Just to clarify. The idea is that count can be used for userspace to check synchronisation with the kernel data. User space doesn't use it because we can check the high/low of the timer instead, but I think the idea is that it could be useful in the future.